README.md
![cymru-api logo](https://raw.githubusercontent.com/blacktop/team-cymru-api/master/doc/logo.png)

team-cymru-api
=========
[![PyPI version](https://badge.fury.io/py/team-cymru-api.svg)](http://badge.fury.io/py/team-cymru-api) [![Build Status](https://travis-ci.org/blacktop/team-cymru-api.svg?branch=master)](https://travis-ci.org/blacktop/team-cymru-api) [![Downloads](https://pypip.in/download/team-cymru-api/badge.png)](https://pypi.python.org/pypi/team-cymru-api/) [![Support blacktop via Gittip](http://img.shields.io/gittip/blacktop.svg)](https://www.gittip.com/blacktop/)

Team Cymru - Malware Hash Registry API

https://www.team-cymru.org/Services/

Installation
-----------

    $ pip install team-cymru-api


Usage
-----
```python
import json
from team_cymru_api import TeamCymruApi

team_cymru = TeamCymruApi()

response =  team_cymru.get_cymru('039ea049f6d0f36f55ec064b3b371c46')
print json.dumps(response, sort_keys=False, indent=4)
```

#### Output:
```json
{
    "last_seen_utc": "2014-01-06T22:34:57Z",
    "response_code": 200,
    "detected": "86"
}
```

Testing
-------

To run the tests:

    $ ./tests

Contributing
------------

1. Fork it.
2. Create a branch (`git checkout -b my_team_cymru_api`)
3. Commit your changes (`git commit -am "Added Something Cool"`)
4. Push to the branch (`git push origin my_team_cymru_api`)
5. Open a [Pull Request](https://github.com/blacktop/team-cymru-api/pulls)
6. Wait for me to figure out what the heck a pull request is...


HISTORY.rst
.. :changelog:

Release History
---------------

1.0.4 (2014-05-18)
------------------

**Fixes**

- Fixing error in README

1.0.2 (2014-05-18)
------------------

**API Changes**

- Changing folder structure so when people import it it is not dumb :(

1.0.1 (2014-04-14)
------------------

**Bugfixes**

- Trying to fix setup.py for deploying to PYPI.


team_cymru_api/test/__init__.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

__author__ = 'Josh Maine'


team_cymru/test/test_team_cymru.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
__author__ = 'Josh Maine'
__version__ = '1.0.1'
__license__ = 'GPLv3'

from unittest import TestCase
from team_cymru.team_cymru_api import TeamCymruApi
import json

class InitTests(TestCase):
    def test_hash_found(self):
        myteam = TeamCymruApi()

        try:
            print json.dumps(myteam.get_cymru('039ea049f6d0f36f55ec064b3b371c46'), sort_keys=True, indent=4)
        except Exception as e:
            self.fail(e)

    def test_hash_not_found(self):
        myteam = TeamCymruApi()

        try:
            print json.dumps(myteam.get_cymru('5e28284f9b5f9097640d58a73d38ad4c'), sort_keys=True, indent=4)
        except Exception as e:
            self.fail(e)

    def test_hash_bad_input(self):
        myteam = TeamCymruApi()

        try:
            print json.dumps(myteam.get_cymru(False), sort_keys=True, indent=4)
        except Exception as e:
            self.fail(e)

team_cymru_api/test/test_team_cymru.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
__author__ = 'Josh Maine'
__version__ = '1.0.1'
__license__ = 'GPLv3'

import json
from unittest import TestCase

from team_cymru_api import TeamCymruApi


class InitTests(TestCase):
    def test_hash_found(self):
        myteam = TeamCymruApi()

        try:
            print json.dumps(myteam.get_cymru('039ea049f6d0f36f55ec064b3b371c46')
                                                                ,sort_keys=True
                                                                ,indent=4)
        except Exception as e:
            self.fail(e)

    def test_hash_not_found(self):
        myteam = TeamCymruApi()

        try:
            print json.dumps(myteam.get_cymru('5e28284f9b5f9097640d58a73d38ad4c')
                                                                ,sort_keys=True
                                                                ,indent=4)
        except Exception as e:
            self.fail(e)

    def test_hash_bad_input(self):
        myteam = TeamCymruApi()

        try:
            print json.dumps(myteam.get_cymru(False), sort_keys=True, indent=4)
        except Exception as e:
            self.fail(e)


team_cymru/api.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

"""
team-cymru-api
~~~~~~~~~~~~~~

This module implements the Team Cymru API.

:copyright: (c) 2014 by Josh "blacktop" Maine.
:license: GPLv3, see LICENSE for more details.
"""

import re
import socket
from time import gmtime, strftime


class TeamCymruApi():
    def __init__(self):
        self.cymru = 'hash.cymru.com'

    def get_cymru(self, this_hash):
        """ Return Team Cymru Malware Hash Database results.

        The Malware Hash Registry (MHR) project is a look-up service similar to
        the Team Cymru IP address to ASN mapping project. This project differs
        however, in that you can query our service for a computed MD5 or SHA-1
        hash of a file and, if it is malware and we know about it, we return
        the last time we've seen it along with an approximate anti-virus
        detection percentage.

        :param this_hash: Can be a md5 or sha1 hash.
        :return: result dictionary or socket error

        Example Output::

            {
                'detected': '86',
                'last_seen': '01-06-2014T22:34:57Z'
            }

        source: http://code.google.com/p/malwarecookbook/
        site : http://www.team-cymru.org/Services/MHR/
        """
        if self.__vaildate_input(this_hash):
            request = '%s\r\n' % this_hash
            s = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
            try:
                s.connect((self.cymru, 43))
                s.send('begin\r\n')
                s.recv(1024)
                s.send(request)
                response = s.recv(1024)
                s.send('end\r\n')
                s.close()
                if len(response) > 0:
                    resp_re = re.compile('\S+ (\d+) (\S+)')
                    match = resp_re.match(response)
                    if 'NO_DATA' in match.group(2):
                        return dict(last_seen_utc=strftime("%Y-%m-%dT%H:%M:%SZ",
                                                   gmtime(int(match.group(1)))),
                                                   detected=match.group(2),
                                                   response_code=404)
                    else:
                        return dict(last_seen_utc=strftime("%Y-%m-%dT%H:%M:%SZ",
                                                   gmtime(int(match.group(1)))),
                                                   detected=match.group(2),
                                                   response_code=200)
            except socket.error:
                return dict(error='socket error', response_code=500)
        else:
            return dict(error='Bad Input', response_code=400)

    def __vaildate_input(self, user_input):
        """ Validate File Hash Input.

        :param user_input: Input string
        :return: str Single hash string
        """
        #: Check that input is a string
        #: Check that it is at least as long as a md5 hash
        if isinstance(user_input, basestring) and len(user_input) >= 32:
            found_md5s = self.__extract_md5_from_hash_list(user_input)
            found_sha1s = self.__extract_sha1_from_hash_list(user_input)
            #: Check that hash list is not empty or mixed
            if found_md5s and not found_sha1s:
                return found_md5s
            elif found_sha1s and not found_md5s:
                return found_sha1s
            else:
                return False
        return False

    @staticmethod
    def __extract_md5_from_hash_list(data):
        """ Extract all md5 hashes from a string.

        :param data: Input string
        :return: List of md5 hashes or single md5 hash
        """
        found = re.findall(r'(?i)(?<![a-zA-Z0-9])[a-fA-F0-9]{32}(?![a-zA-Z0-9])', data)
        hash_list = list(set(found))
        if len(hash_list) == 1:
            return hash_list[0]
        else:
            return hash_list

    @staticmethod
    def __extract_sha1_from_hash_list(data):
        """ Extract all sha1 hashes from a string.

        :param data: Input string
        :return: List of sha1 hashes or single sha1 hash
        """
        found = re.findall(r'(?i)(?<![a-zA-Z0-9])[a-fA-F0-9]{40}(?![a-zA-Z0-9])', data)
        hash_list = list(set(found))
        if len(hash_list) == 1:
            return hash_list[0]
        else:
            return hash_list


team_cymru/team_cymru_api.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
__author__ = 'Josh Maine'
__version__ = '1.0.1'
__license__ = 'GPLv3'

import re
import socket
from time import strftime, gmtime


class TeamCymruApi():
    def __init__(self):
        self.cymru = 'hash.cymru.com'

    def get_cymru(self, this_hash):
        """ Return Team Cymru Malware Hash Database results.

        The Malware Hash Registry (MHR) project is a look-up service similar to
        the Team Cymru IP address to ASN mapping project. This project differs
        however, in that you can query our service for a computed MD5 or SHA-1
        hash of a file and, if it is malware and we know about it, we return the
        last time we've seen it along with an approximate anti-virus detection percentage.

        :param this_hash: Can be a md5 or sha1 hash.
        :return: result dictionary or socket error

        Example Output::

            {
                'detected': '86',
                'last_seen': '01-06-2014T22:34:57Z'
            }

        source: http://code.google.com/p/malwarecookbook/
        site : http://www.team-cymru.org/Services/MHR/
        """
        if self.__vaildate_input(this_hash):
            request = '%s\r\n' % this_hash
            s = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
            try:
                s.connect((self.cymru, 43))
                s.send('begin\r\n')
                s.recv(1024)
                s.send(request)
                response = s.recv(1024)
                s.send('end\r\n')
                s.close()
                if len(response) > 0:
                    resp_re = re.compile('\S+ (\d+) (\S+)')
                    match = resp_re.match(response)
                    if 'NO_DATA' in match.group(2):
                        return dict(last_seen_utc=strftime("%Y-%m-%dT%H:%M:%SZ", gmtime(int(match.group(1)))),
                                    detected=match.group(2), response_code=404)
                    else:
                        return dict(last_seen_utc=strftime("%Y-%m-%dT%H:%M:%SZ", gmtime(int(match.group(1)))),
                                    detected=match.group(2), response_code=200)
            except socket.error:
                return dict(error='socket error', response_code=500)
        else:
            return dict(error='Bad Input', response_code=400)

    def __vaildate_input(self, user_input):
        """ Validate File Hash Input.

        :param user_input: Input string
        :return: str Single hash string
        """
        #: Check that input is a string
        #: Check that it is at least as long as a md5 hash
        if isinstance(user_input, basestring) and len(user_input) >= 32:
            found_md5s = self.__extract_md5_from_hash_list(user_input)
            found_sha1s = self.__extract_sha1_from_hash_list(user_input)
            #: Check that hash list is not empty or mixed
            if found_md5s and not found_sha1s:
                return found_md5s
            elif found_sha1s and not found_md5s:
                return found_sha1s
            else:
                return False
        return False

    @staticmethod
    def __extract_md5_from_hash_list(data):
        """ Extract all md5 hashes from a string.

        :param data: Input string
        :return: List of md5 hashes or single md5 hash
        """
        found_md5_hashes = re.findall(r'(?i)(?<![a-zA-Z0-9])[a-fA-F0-9]{32}(?![a-zA-Z0-9])', data)
        hash_list = list(set(found_md5_hashes))
        if len(hash_list) == 1:
            return hash_list[0]
        else:
            return hash_list

    @staticmethod
    def __extract_sha1_from_hash_list(data):
        """ Extract all sha1 hashes from a string.

        :param data: Input string
        :return: List of sha1 hashes or single sha1 hash
        """
        found_sha1_hashes = re.findall(r'(?i)(?<![a-zA-Z0-9])[a-fA-F0-9]{40}(?![a-zA-Z0-9])', data)
        hash_list = list(set(found_sha1_hashes))
        if len(hash_list) == 1:
            return hash_list[0]
        else:
            return hash_list